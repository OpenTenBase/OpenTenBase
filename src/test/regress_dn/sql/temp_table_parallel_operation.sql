\c regression_ora

set enable_parallel_insert to on;
set enable_parallel_update to on;
set parallel_setup_cost to 0;
set parallel_tuple_cost to 0;
set min_parallel_table_scan_size to 0;
set max_parallel_workers_per_gather=2;

-- create normal table
drop table if exists baseTable;
create table baseTable(col1 int primary key,
                       col2 varchar2(200),
                       col3 int,
                       parent int,
                       varchar_col VARCHAR2(255),
                       nvarchar_col NVARCHAR2(255),
                       char_col CHAR(10),
                       nchar_col NCHAR(10),
                       number_col NUMBER,
                       integer_col INTEGER,
                       float_col FLOAT,
                       date_col DATE
                      );

-- insert data to normal table;
insert into baseTable
select n,
       'col2_'||n,
       n,
       n%100+1,
       'Test VARCHAR2 - ',
       N'Test NVARCHAR2',
       'Test CHAR',
       N'Test NCHAR',
       123.45,
       42,
       3.1415,
       TO_DATE('2024-03-08', 'YYYY-MM-DD')
from generate_series(1,1000) n;

-- global temp table operation tests
drop table if exists gttTable;
CREATE GLOBAL TEMPORARY TABLE gttTable (col1 int primary key,
                                        col2 varchar2(200),
                                        col3 int,
                                        parent int,
                                        varchar_col VARCHAR2(255),
                                        nvarchar_col NVARCHAR2(255),
                                        char_col CHAR(10),
                                        nchar_col NCHAR(10),
                                        number_col NUMBER,
                                        integer_col INTEGER,
                                        float_col FLOAT,
                                        date_col DATE
                                       ) ON COMMIT delete ROWS;

begin;

explain (costs off)  insert into gttTable select * from baseTable;
insert into gttTable select * from baseTable;

explain (costs off) select count(*)  from gttTable;
select count(*)  from gttTable;

explain (costs off) select col3 from gttTable where col3 < 100;

explain (costs off) update gttTable set col2 = col2 || ' - 2' where col1 < 1001;
update gttTable set col2 = col2 || ' - 2' where col1 < 1001;
select col2 from gttTable  where col1 <= 5 ;

commit ;
select count(*)  from gttTable;
drop table gttTable;

-- temp table parallel operation tests
drop table if exists tempTable;
CREATE TEMPORARY TABLE tempTable (col1 int primary key,
                                  col2 varchar2(200),
                                  col3 int,
                                  parent int,
                                  varchar_col VARCHAR2(255),
                                  nvarchar_col NVARCHAR2(255),
                                  char_col CHAR(10),
                                  nchar_col NCHAR(10),
                                  number_col NUMBER,
                                  integer_col INTEGER,
                                  float_col FLOAT,
                                  date_col DATE
                                 )     ON COMMIT delete ROWS;

begin;

explain (costs off)  insert into tempTable select * from baseTable;
insert into tempTable select * from baseTable;

explain (costs off) select count(*)  from tempTable;
select count(*)  from tempTable;

explain (costs off) select col3 from tempTable where col3 < 100;

explain (costs off) update tempTable set col2 = col2 || ' - 2' where col1 < 1001;
update tempTable set col2 = col2 || ' - 2' where col1 < 1001;
select col2 from tempTable  where col1 <= 5 ;

commit ;
select count(*)  from tempTable;
drop table tempTable;

