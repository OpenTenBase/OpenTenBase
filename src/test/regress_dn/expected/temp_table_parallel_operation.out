\c regression_ora
set enable_parallel_insert to on;
set enable_parallel_update to on;
set parallel_setup_cost to 0;
set parallel_tuple_cost to 0;
set min_parallel_table_scan_size to 0;
set max_parallel_workers_per_gather=2;
-- create normal table
drop table if exists baseTable;
NOTICE:  table "BASETABLE" does not exist, skipping
create table baseTable(col1 int primary key,
                       col2 varchar2(200),
                       col3 int,
                       parent int,
                       varchar_col VARCHAR2(255),
                       nvarchar_col NVARCHAR2(255),
                       char_col CHAR(10),
                       nchar_col NCHAR(10),
                       number_col NUMBER,
                       integer_col INTEGER,
                       float_col FLOAT,
                       date_col DATE
                      );
-- insert data to normal table;
insert into baseTable
select n,
       'col2_'||n,
       n,
       n%100+1,
       'Test VARCHAR2 - ',
       N'Test NVARCHAR2',
       'Test CHAR',
       N'Test NCHAR',
       123.45,
       42,
       3.1415,
       TO_DATE('2024-03-08', 'YYYY-MM-DD')
from generate_series(1,1000) n;
-- global temp table operation tests
drop table if exists gttTable;
NOTICE:  table "GTTTABLE" does not exist, skipping
CREATE GLOBAL TEMPORARY TABLE gttTable (col1 int primary key,
                                        col2 varchar2(200),
                                        col3 int,
                                        parent int,
                                        varchar_col VARCHAR2(255),
                                        nvarchar_col NVARCHAR2(255),
                                        char_col CHAR(10),
                                        nchar_col NCHAR(10),
                                        number_col NUMBER,
                                        integer_col INTEGER,
                                        float_col FLOAT,
                                        date_col DATE
                                       ) ON COMMIT delete ROWS;
WARNING:  GLOBAL is deprecated in temporary table creation
LINE 1: CREATE GLOBAL TEMPORARY TABLE gttTable (col1 int primary key...
               ^
begin;
explain (costs off)  insert into gttTable select * from baseTable;
                 QUERY PLAN                 
--------------------------------------------
 Gather
   Workers Planned: 2
   ->  Parallel Insert on GTTTABLE
         ->  Parallel Seq Scan on BASETABLE
(4 rows)

insert into gttTable select * from baseTable;
explain (costs off) select count(*)  from gttTable;
                   QUERY PLAN                    
-------------------------------------------------
 Finalize Aggregate
   ->  Gather
         Workers Planned: 2
         ->  Partial Aggregate
               ->  Parallel Seq Scan on GTTTABLE
(5 rows)

select count(*)  from gttTable;
 COUNT 
-------
  1000
(1 row)

explain (costs off) select col3 from gttTable where col3 < 100;
             QUERY PLAN              
-------------------------------------
 Gather
   Workers Planned: 2
   ->  Parallel Seq Scan on GTTTABLE
         Filter: (COL3 < 100)
(4 rows)

explain (costs off) update gttTable set col2 = col2 || ' - 2' where col1 < 1001;
                QUERY PLAN                 
-------------------------------------------
 Gather
   Workers Planned: 2
   ->  Parallel Update on GTTTABLE
         ->  Parallel Seq Scan on GTTTABLE
               Filter: (COL1 < 1001)
(5 rows)

update gttTable set col2 = col2 || ' - 2' where col1 < 1001;
select col2 from gttTable  where col1 <= 5 ;
    COL2    
------------
 col2_1 - 2
 col2_2 - 2
 col2_3 - 2
 col2_4 - 2
 col2_5 - 2
(5 rows)

commit ;
select count(*)  from gttTable;
 COUNT 
-------
     0
(1 row)

drop table gttTable;
-- temp table parallel operation tests
drop table if exists tempTable;
NOTICE:  table "TEMPTABLE" does not exist, skipping
CREATE TEMPORARY TABLE tempTable (col1 int primary key,
                                  col2 varchar2(200),
                                  col3 int,
                                  parent int,
                                  varchar_col VARCHAR2(255),
                                  nvarchar_col NVARCHAR2(255),
                                  char_col CHAR(10),
                                  nchar_col NCHAR(10),
                                  number_col NUMBER,
                                  integer_col INTEGER,
                                  float_col FLOAT,
                                  date_col DATE
                                 )     ON COMMIT delete ROWS;
begin;
explain (costs off)  insert into tempTable select * from baseTable;
                 QUERY PLAN                 
--------------------------------------------
 Gather
   Workers Planned: 2
   ->  Parallel Insert on TEMPTABLE
         ->  Parallel Seq Scan on BASETABLE
(4 rows)

insert into tempTable select * from baseTable;
explain (costs off) select count(*)  from tempTable;
                    QUERY PLAN                    
--------------------------------------------------
 Finalize Aggregate
   ->  Gather
         Workers Planned: 2
         ->  Partial Aggregate
               ->  Parallel Seq Scan on TEMPTABLE
(5 rows)

select count(*)  from tempTable;
 COUNT 
-------
  1000
(1 row)

explain (costs off) select col3 from tempTable where col3 < 100;
              QUERY PLAN              
--------------------------------------
 Gather
   Workers Planned: 2
   ->  Parallel Seq Scan on TEMPTABLE
         Filter: (COL3 < 100)
(4 rows)

explain (costs off) update tempTable set col2 = col2 || ' - 2' where col1 < 1001;
                 QUERY PLAN                 
--------------------------------------------
 Gather
   Workers Planned: 2
   ->  Parallel Update on TEMPTABLE
         ->  Parallel Seq Scan on TEMPTABLE
               Filter: (COL1 < 1001)
(5 rows)

update tempTable set col2 = col2 || ' - 2' where col1 < 1001;
select col2 from tempTable  where col1 <= 5 ;
    COL2    
------------
 col2_1 - 2
 col2_2 - 2
 col2_3 - 2
 col2_4 - 2
 col2_5 - 2
(5 rows)

commit ;
select count(*)  from tempTable;
 COUNT 
-------
     0
(1 row)

drop table tempTable;
